docs-preview/src/extension.ts

```diff
@@ -35,12 +35,217 @@ export function activate(context: ExtensionContext) {
+                        }
+                    },
+                })
+                .use(container_plugin, "column", {
+                    marker: ":",
+                    validate(params) {
+                        return params.trim().match(/column(\s+span="([1-9]+)?")?:::/g) || params.trim().match(/column-end:::/g);
+                    },
+                    render(tokens, idx) {
+                        const RE = /column((\s+)span="([1-9]+)?")?:::/g;
+                        const start = RE.exec(tokens[idx].info.trim());
+                        if (start) {
+                            if (start[3]) {
+                                return `<div class='column span${start[3]}'>`;
+                            } else {
+                                // opening tag
+                                return "<div class='column'>";
+                            }
+                        } else {
+                            // closing tag
+                            return "</div>";
+                        }
+                    },
+                })
+        },
     };
 }
 
 // this method is called when your extension is deactivated
 export function deactivate() {
     output.appendLine("Deactivating extension.");
 }
+
+export function isMarkdownFile(document: TextDocument) {
+    return document.languageId === "markdown"; // prevent processing of own documents
+}
+
+export function container_plugin(md, name, options) {
+    options = options || {};
+    let min_markers = 3,
+        marker_str = options.marker,
+        marker_len = marker_str.length,
+        marker_char = marker_str.charCodeAt(0),
+        validate = options.validate,
+        render = options.render;
+
+    function container(state, startLine, endLine, silent) {
+        let pos, nextLine, marker_count, markup, params, token,
+            old_parent, old_line_max,
+            auto_closed = false,
+            start = state.bMarks[startLine] + state.tShift[startLine],
+            max = state.eMarks[startLine];
+
+        // Check out the first character quickly,
+        // this should filter out most of non-containers
+        //
+        if (marker_char !== state.src.charCodeAt(start)) { return false; }
+
+        // Check out the rest of the marker string
+        //
+        for (pos = start; pos <= max; pos++) {
+            if (":" !== state.src[pos]) {
+                break;
+            }
+        }
+        // marker_count = Math.floor((pos - start) / marker_len);
+        // if (marker_count < min_markers) { return false; }
+
+        pos -= (pos - start) % marker_len;
+
+        markup = state.src.slice(start, pos);
+        params = state.src.slice(pos, max);
+        if (!validate(params)) { return false; }
+
+        // Since start is found, we can report success here in validation mode
+        //
+        if (silent) { return true; }
+
+        // Search for the end of the block
+        //
+        nextLine = startLine;
+
+        for (; ;) {
```

Comment:
Wow... I've never seen this way of using a for loop... is this similar to:
`while(nextLine < endline){...`

?